Use bootstrap4

OGAWAHirofumi opened the pull request:
This migrate bootstrap3 to bootstrap4. With this, IMO, becomes better views (but im not a web developer).

however, testing and checking UI is hard (no auto test, and people has own favorite), so this PR would be needed carefully review more or less.

ThomasWaldmann commented:
Can you do a **minimal** change please to upgrade to bs4?

OGAWAHirofumi commented:
> Can you do a **minimal** change please to upgrade to bs4?

minimal means, doesn't care bad view by convert? e.g. looks like some css class (e.g. page-header) was removed from bootstrap4.

ThomasWaldmann commented:
Well, I see some changes that do not look minimal, like:

Size changes  `sm -> md` (small to medium)?

Colours are being added here and there, why?

Some stuff looks like new features, like:
- Use shorter name for "Create List Item"
- Use full width for navbar like bootstrap4 examples

In a pull request that upgrades bootstrap to a new version, there should be only changes that make the UI work with bootstrap 4.

Regression fixes related to that upgrade are also ok (so it does not look obviously bad / wrong **if** it looked ok with bs3).

But other than that, there should be no other changes **within this PR**.

**Improvements on looks and usability are welcome in general, but should be separate.**

OGAWAHirofumi commented:
> Well, I see some changes that do not look minimal, like:
> 
> Size changes `sm -> md` (small to medium)?

col-<size> were changed in bootstrap4. like xs => default, sm => md.

> Colours are being added here and there, why?

<component>-default was removed.

> Some stuff looks like new features, like:
> 
>     * Use shorter name for "Create List Item"
> 
>     * Use full width for navbar like bootstrap4 examples
> 
> 
> In a pull request that upgrades bootstrap to a new version, there should be only changes that make the UI work with bootstrap 4.
> 
> Regression fixes related to that upgrade are also ok (so it does not look obviously bad / wrong **if** it looked ok with bs3).
> 
> But other than that, there should be no other changes **within this PR**.
> 
> **Improvements on looks and usability are welcome in general, but should be separate.**

ah, it was why separated patches (can revert easily). but you wanted separated **PR** somewhat. ok, will separate.

OGAWAHirofumi commented:
Added screen shot script to see UI changes.

[screenshots-before.tar.gz](https://github.com/bepasty/bepasty-server/files/5500373/screenshots-before.tar.gz)
[screenshots-after.tar.gz](https://github.com/bepasty/bepasty-server/files/5500375/screenshots-after.tar.gz)

screenshots-before == bootstrap3
screenshots-after == bootstrap4

those should make easier to review.

with result of screenshots, adjusted button wrap in display view, and tweaked QR view to be more bootstrap3

ThomasWaldmann commented on src/bepasty/tests/screenshots.py:
```diff
@@ -0,0 +1,203 @@
+        try:
+            self.browser.find_element_by_xpath("//input[@value='Logout']")
+        except NoSuchElementException:
+            raise ValueError("Can't login!!! Create a user 'foo' with the permissions"
+                             "'read' and 'create' in your PERMISSIONS in the config")
```
not "user", it is a "secret".

So it becomes:

Can't login! Please edit your config, go to PERMISSIONS setting and add a new secret 'screenshots' with 'read' and 'create' permissions.

ThomasWaldmann commented:
Looks good! One minor issue to fix and it's ready to merge!

ThomasWaldmann commented:
Guess I need to update the bootstrap 4 xstatic package ...

OGAWAHirofumi commented:
changed

- tweaked screenshots.py error message to say for 'foo' and all permissions
- use xstatic-bootstrap>=4.0.0.0,<5.0.0.0
- screenshots.tar.gz with bootstrap 4.5.1 (Create List Item button wrap now, this will fix separatedly)

[screenshots.tar.gz](https://github.com/bepasty/bepasty-server/files/5505870/screenshots.tar.gz)

OGAWAHirofumi commented:
tested with latest xstatic stuff

only noticed change looks like `<pre>` color in pygment (background color for line number)

[screenshots.tar.gz](https://github.com/bepasty/bepasty-server/files/5505930/screenshots.tar.gz)

OGAWAHirofumi commented:
tested with latest xstatic stuff. only noticed change looks like `<pre>` color in pygment (background color for line number)

[screenshots.tar.gz](https://github.com/bepasty/bepasty-server/files/5505948/screenshots.tar.gz)

changed

- improved screenshots.py to take screen shot for bootbox

previous

- tweaked screenshots.py error message to say for 'foo' and all permissions
- use xstatic-bootstrap>=4.0.0.0,<5.0.0.0
- screenshots.tar.gz with bootstrap 4.5.1 (Create List Item button wrap now, this will fix separatedly)

OGAWAHirofumi commented:
tested with latest xstatic stuff. only noticed change looks like `<pre>` color in pygment (background color for line number)

[screenshots.tar.gz](https://github.com/bepasty/bepasty-server/files/5506071/screenshots.tar.gz)

changed

- tweaked screen height for screenshots.py
- improved screenshots.py to take screen shot for bootbox
- added line highlight for display view screen shot

previous

- tweaked screenshots.py error message to say for 'foo' and all permissions
- use xstatic-bootstrap>=4.0.0.0,<5.0.0.0

ThomasWaldmann commented on setup.py:
```diff
@@ -43,7 +43,7 @@
         'XStatic-asciinema-player',
         'xstatic-bootbox',
-        'xstatic-bootstrap<4.0.0.0',
+        'xstatic-bootstrap>=4.0.0.0,<5.0.0',
```
please use 4 digits for both version numbers

OGAWAHirofumi replied:
ouch, will fix

ThomasWaldmann commented on src/bepasty/tests/screenshots.py:
```diff
@@ -0,0 +1,203 @@
+        try:
+            self.browser.find_element_by_xpath("//input[@value='Logout']")
+        except NoSuchElementException:
+            raise ValueError("Can't login! Please edit your config, go to PERMISSIONS setting "
+                             "and add a new secret 'foo' with all permissions.")
```
can you use a better secret than 'foo' here? maybe 'for-screenshots'?

OGAWAHirofumi replied:
'foo' is default entry in config.py. and it is what test_website.py uses too.
changing from 'foo', just will make inconvenient to test.

ThomasWaldmann replied:
oh, ok. so keep it for now.

OGAWAHirofumi commented:
tested with latest xstatic stuff. only noticed change looks like `<pre>` color in pygment (background color for line number)

[screenshots.tar.gz](https://github.com/bepasty/bepasty-server/files/5506071/screenshots.tar.gz)

changed

- use 5.0.0.0

previous

- tweaked screen height for screenshots.py
- improved screenshots.py to take screen shot for bootbox
- added line highlight for display view screen shot
- tweaked screenshots.py error message to say for 'foo' and all permissions
- use xstatic-bootstrap>=4.0.0.0,<5.0.0.0
- screenshots.tar.gz with bootstrap 4.5.1 (Create List Item button wrap now, this will fix separatedly)

ThomasWaldmann commented:
OK, looks like this is finished! Will merge...
